engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypePropagation.java

```diff
@@ -164,6 +175,45 @@ TypeRepresentation tryInferringType(
     return ascribedType != null ? ascribedType : inferredType;
   }
 
+  private boolean canBeDiscarded(TypeRepresentation type) {
+    if (type == null || type instanceof TypeRepresentation.TopType) {
+      // If the type is unknown then we allow it to be discarded as we don't know enough yet
+      return true;
+    }
+
+    if (type.equals(BuiltinTypes.NOTHING)) {
+      // Nothing is the type that side-effectful functions should return - it is most often meant to
+      // be discarded.
+      return true;
+    }
+
+    // Sometimes Nothing is inferred as `type Nothing`. Maybe we should fix it, but for now:
```

Comment:
`Nothing` is (Java) `==` with `type Nothing` in the runtime. In fact all singletons are equal with its type.

Reply:
Perhaps I need to adapt the `TypeRepresentation` structure to more closely match this for singleton types.